[INFRA] Convert entity table to yaml (#475)
bids-standard/bids-specification: CONTRIBUTING.md

<img align="center" src="https://i.imgur.com/s8yELfK.png" alt="Example-Contribution" width="800"/>

## Updating the schema

Portions of the BIDS specification are defined using YAML files, in order to make the specification machine-readable.
Currently, the only portion of the specification that relies on this schema is the Entity Table, but any changes to the specification should be mirrored in the schema.

### The format of the schema

The schema reflects the files and objects in the specification, as well as associations between these objects.

> Here is a list of the files and subfolders of the schema, with a description for each:

Reviewer: What is the order here?  At first I thought it was alphabetical, but it is not, and the first item listed is (as far as I understand) one of the least important YAML files.

Can I suggest reordering so that they are in some order that reflects how you might have to sift through them if you are a BIDS contributor making a change.

Then maybe change this:

```suggestion
Here is a list of the files and subfolders of the schema, roughly in order of importance:
```

(No need to say that you're going to give a description... that's clear).

- `associated_data.yaml`:
    Folders that contain files falling outside the jurisdiction of the specification, but which are commonly contained within the same folder as a BIDS dataset.

>     The schema specifies which folders are accepted and whether they are required or optional.

Reviewer: I might just be being dense here, but are are folders "outside the jurisdiction" of the spec but which still can be considered "required"?  Isn't that a contradiction?

Author: That's a good point. I included a `required` field for the items in this file mostly to match my convention in other files, but we can assume anything in the file is not required. Rather than abandon that right now, I'll try to make the text a little less confusing. If that doesn't resolve it, I can drop the `required` field in the file.

- `top_level_files.yaml`:
    Modality-agnostic files stored at the top level of a BIDS dataset.
    The schema specifies whether these files are required or optional, as well as acceptable extensions for each.
- `modalities.yaml`:

>     Modalities supported by the specification, along with a list of associated datatypes.

Reviewer: AFAIK, "datatype" is not a thing in the spec. (e.g. I just searched the spec and found no reference to "datatype".  Can you give an explanation here or, at the top of the list, what is meant by "datatype"?

Author: Data types fall under the common principles. They correspond to the folders (`anat`, `func`, `meg`, etc.). I can split the word in the text to make it clearer, and maybe link back to the relevant part of common principles?

Reviewer: Ah!  "data type" two words!  Yes, of course.

- `entities.yaml`:
    A list of entities (key/value pairs in folder and filenames) with associated descriptions and formatting rules.
    The order of the entities in the file determines the order in which entities must appear in filenames.

> - `datatypes/*.yaml`:

Reviewer: Oh... "datatypes".  This seems important.

Can this be the first item in the list?  And, also, from this description, I still don't see what a "datatype" is? I can only guess that you *don't* mean binary data representation (char, float, double), nor text encoding scheme, nor PET vs MRI... maybe `nii.gz` vs `csv`?  

Anyway, you see you really need some example to clarify this.

    Datatypes supported by the specification.
    Each datatype may support many suffixes.
    These suffixes are divided into groups based on what extensions and entities are allowed for each.

> - `auxdatatypes/*.yaml`:

Reviewer: Auxiliary data types also seems like it needs some examples... can't quite guess what this is supposed to be.

Author: This, and your other comments, should be addressed now. Thanks!

Reviewer: Super!  Looks much clearer now!

    Auxiliary (not directly imaging or data-containing) datatypes supported by the specification.
    Each auxiliary datatype is associated with a set of datatypes, and these auxiliary datatypes are grouped based on what datatypes, extensions, and entities are allowed for each.

### Making a change to the schema

#### 1. Ensure that changes to the specification are matched in the schema

The schema formalizes the rules described in the specification text, so you must ensure that any changes which impact the rules of the specification (including, but not limited to, adding new entities, suffixes, datatypes, modalities, etc.) are reflected in the schema as well.

#### 2. Generate an updated entity table before pushing your changes

Run the Python script `tools/bids_schema.py`:

```bash
cd tools/

> python bids_schema.py entity ../src/schema/ ../src/99-appendices/04-entity-table.md

Reviewer: FYI, just can confirm that I tried running this command on my Mac... had to get the right version of Python and install a few packages, but then it ran no problem.  The created MD table looks fine.

Reviewer 2: FWIW, I see such need for manual run if the tool as temporary. Filed https://github.com/bids-standard/bids-specification/issues/544 to automate, and will look into it unless someone beats me to it

    The schema specifies whether these files are required or optional, as well as acceptable extensions for each.
- `modalities.yaml`:
    Modalities supported by the specification, along with a list of associated data types.
    Modalities are not reflected directly in the BIDS structure, but data types are modality-specific.
- `associated_data.yaml`:
    Folders that are commonly contained within the same folder as a BIDS dataset, but which do not follow the BIDS structure internally, such as `code` or `sourcedata`.
    The schema specifies which folders are accepted and whether they are required or optional.

### Making a change to the schema

#### 1. Ensure that changes to the specification are matched in the schema

The schema formalizes the rules described in the specification text, so you must ensure that any changes which impact the rules of the specification (including, but not limited to, adding new entities, suffixes, datatypes, modalities, etc.) are reflected in the schema as well.

#### 2. Generate an updated entity table before pushing your changes

Run the Python script `tools/bids_schema.py`:

```bash

> cd tools/
> python bids_schema.py entity ../src/schema/ ../src/99-appendices/04-entity-table.md

Reviewer 3: `chdir`ing and passing `../`paths seems odd.

```suggestion
python tools/bids_schema.py entity src/schema/ src/99-appendices/04-entity-table.md
```
